addons/bus/websocket.py

```diff
@@ -223,6 +224,31 @@ class Websocket:
     # or received within CONNECTION_TIMEOUT - 15 seconds.
     CONNECTION_TIMEOUT = 60
     INACTIVITY_TIMEOUT = CONNECTION_TIMEOUT - 15
+    # How much time (in second) the history of last dispatched notifications is
+    # kept in memory for this websocket.
+    # To avoid duplicate notifications, we fetch them based on their ids.
+    # However during parallel transactions, ids are assigned immediately (when
+    # they are requested), but the notifications are dispatched at the time of
+    # the commit. This means lower id notifications might be dispatched after
+    # higher id notifications.
+    # Simply incrementing the last id is sufficient to guarantee no duplicates,
+    # but it is not sufficient to guarantee all notifications are dispatched,
+    # and in particular not sufficient for those with a lower id coming after a
+    # higher id was dispatched.
+    # To solve the issue of missed notifications, the lowest id, stored in
+    # ``_last_notif_sent_id``, is held back by a few seconds to give time for
+    # concurrent transactions to finish. To avoid dispatching duplicate
+    # notifications, the history of already dispatched notifications during this
+    # period is kept in memory in ``_notif_history`` and the corresponding
+    # notifications are discarded from subsequent dispatching even if their id
+    # is higher than ``_last_notif_sent_id``.
+    # In practice, what is important functionally is the time between the create
+    # of the notification and the commit of the transaction in business code.
+    # If this time exceeds this threshold, the notification will never be
+    # dispatched if the target user receive any other notification in the
+    # meantime. Transactions known to be long should therefore create their
+    # notifications at the end, as close as possible to their commit.
+    MAX_NOTIFICATION_HISTORY_SEC = 10
```

Comment:
10 seconds looks a lot when this works flawlessly... But could be too close to slow connectivity practical scenario. I would personally put value closer to `TIMEOUT` so in the `30-40`ish values at least.

We should take into account some long connectivity like being physically in Europe but server is in Japan or Australia.

Reply:
The delay introduced here is specifically targeted at the delay between `create()` on `bus.bus` and the end of the respective request/transaction in python.

It's normally not impacted by network or any other consideration than slowness in business code:
- create bus -> sleep longer than the delay as a dummy example (which shouldn't be done)
- or create bus followed by a very slow query or cpu heavy processing (which is indeed not supported, unless the create bus is done _after_ the processing)

Notifications can still be fetched much later in practice. If a user (re)connects its socket after a disconnect, all the notif that are after its last known id are still fetched, as far away in the past as necessary (and as long as it doesn't go above garbage collect threshold of course).

The delay here is "how long do we still wait for lower id notifications when receiving a higher id one", and that disordering of ids that we are waiting for only happens at the transaction/commit level.

Reply:
I'm not against increasing the number if we have resources to spare and want to be a little bit more on the safe side however. Anything up to 60s looks reasonable to me.

Reply:
Oh ok I see much better the idea. Thanks for explanation!

Reply:
It still got me thinking that it would be best to move the clean up of the list after the dispatching instead of before to give a chance for delayed dispatching to catch up with missed ids once list time before cleaning them.